python/ql/lib/semmle/python/security/dataflow/TarSlipCustomizations.qll

```diff
@@ -0,0 +1,141 @@
+      // this tarfile is safe.
+      not this.(DataFlow::CallCfgNode).getArg(0).getALocalSource().asExpr() instanceof StrConst and
+      /* Ignore opens within the tarfile module itself */
+      not this.getLocation().getFile().getBaseName() = "tarfile.py"
+    }
+  }
+
+  /**
+   * A sanitizer based on file name. This because we extract the standard library.
+   *
+   * For efficiency we don't want to track the flow of taint
+   * around the tarfile module.
+   */
+  class ExcludeTarFilePy extends Sanitizer {
+    ExcludeTarFilePy() { this.getLocation().getFile().getBaseName() = "tarfile.py" }
+  }
+
+  /**
+   * A sink capturing method calls to `extractall`.
+   *
+   * For a call to `file.extractall` without arguments, `file` is considered a sink.
+   */
+  class ExtractAllSink extends Sink {
+    ExtractAllSink() {
+      exists(DataFlow::CallCfgNode call |
+        call =
+          API::moduleImport("tarfile")
+              .getMember("open")
+              .getReturn()
+              .getMember("extractall")
+              .getACall() and
+        not exists(call.getArg(_)) and
+        not exists(call.getArgByName(_)) and
+        this = call.(DataFlow::MethodCallNode).getObject()
+      )
+    }
+  }
+
+  /**
+   * An argument to `extract` is considered a sink.
+   */
+  class ExtractSink extends Sink {
+    ExtractSink() {
+      exists(DataFlow::CallCfgNode call |
+        call =
+          API::moduleImport("tarfile").getMember("open").getReturn().getMember("extract").getACall() and
+        this = call.getArg(0)
+      )
+    }
+  }
+
+  /** The `members` argument `extractall` is considered a sink. */
+  class ExtractMembersSink extends Sink {
+    ExtractMembersSink() {
+      exists(DataFlow::CallCfgNode call |
+        call =
+          API::moduleImport("tarfile")
+              .getMember("open")
+              .getReturn()
+              .getMember("extractall")
+              .getACall() and
+        this in [call.getArg(0), call.getArgByName("members")]
+      )
+    }
+  }
+
+  /**
+   * A sanitizer guard heuristic.
+   *
+   * For a "check-like function-name" (matching `"%path"`), `checkPath`,
+   * and a call `checkPath(info.name)`, the variable `info` is considered checked.
+   */
+  class TarFileInfoSanitizer extends SanitizerGuard {
+    ControlFlowNode tarInfo;
+
+    TarFileInfoSanitizer() {
+      exists(CallNode call, AttrNode attr |
+        this = call and
+        // We must test the name of the tar info object.
+        attr = call.getAnArg() and
+        attr.getName() = "name" and
+        attr.getObject() = tarInfo
+      |
+        // Assume that any test with "path" in it is a sanitizer
+        call.getAChild*().(AttrNode).getName().matches("%path")
+        or
+        call.getAChild*().(NameNode).getId().matches("%path")
+      )
+    }
+
+    override predicate checks(ControlFlowNode checked, boolean branch) {
+      checked = tarInfo and
+      branch in [true, false]
```

Comment:
the old code only clears taint on the false edge. I don't see any mention of why this was changed, but the fact that our tests results did not change because of this, hints that our tests could have been better :smile:

Reply:
True, I never got to write this down. I felt that the heuristic "contains `path`" did not warrant an assumption about whether the function tested for good or bad paths...